File: js-port/fs-js/src/funcs/text/join-text-function.js
const { BaseFunction, CallType } = require('../../core/function-base');
const helpers = require('../helpers');
const { FSDataType } = require('../../core/fstypes');

class JoinTextFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'join';
    this.callType = CallType.Dual;
  }

  get maxParameters() {
    return 2;
  }

  evaluate(provider, parameters) {
    const error = helpers.expectParamCount(this.symbol, parameters, this.maxParameters);
    if (error) {
      return error;
    }

    const list = helpers.ensureList(parameters.getParameter(provider, 0));
    if (!list) {
      return helpers.makeError(
        helpers.FsError.ERROR_TYPE_MISMATCH,
        `${this.symbol}: first parameter should be a list`
      );
    }

    const separatorResult = helpers.requireString(this.symbol, parameters.getParameter(provider, 1), 'separator');
    if (!separatorResult.ok) {
      return separatorResult.error;
    }
    const separator = separatorResult.value;

    const parts = [];
    for (const item of list) {
      if (item == null) continue;
      const str = helpers.ensureTyped(item);
      if (helpers.typeOf(str) === FSDataType.Null) continue;
      parts.push(String(helpers.valueOf(str)));
    }

    return helpers.makeValue(FSDataType.String, parts.join(separator));
  }

  parName(index) {
    return index === 0 ? 'List' : index === 1 ? 'Separator' : '';
  }
}

module.exports = {
  JoinTextFunction
};

File: FunscScript/Functions/Text/JoinTextFunction.cs
using FunscScript.Core;
using FunscScript.Model;
using System.Text;

namespace FunscScript.Functions.Text
{
    public class JoinTextFunction : IFsFunction
    {
        public const string SYMBOL = "join";
        public int MaxParsCount => 2;

        public CallType CallType => CallType.Dual;

        public string Symbol => SYMBOL;

        public int Precedence => 100;

        public object Evaluate(IFsDataProvider parent, IParameterList pars)
        {
            if (pars.Count != MaxParsCount)
                throw new Error.TypeMismatchError($"{this.Symbol}: Two parameters expected");

            var par0 = pars.GetParameter(parent, 0);
            var par1 = pars.GetParameter(parent, 1);

            if (par0 == null || par1 == null)
                throw new Error.TypeMismatchError($"{this.Symbol}: List and separator expected as parameters");
            if(!(par0 is FsList list))
               throw new InvalidOperationException($"{this.Symbol}: first parameter should be list");
            if(!(par1 is string separator))
                throw new InvalidOperationException($"{this.Symbol}: second parameter should be string");
            
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < list.Length; i++)
            {
                var item = list[i];
                if (item != null)
                {
                    if (i > 0)
                        sb.Append(separator);
                    sb.Append(item?? "");
                }
            }
            return sb.ToString();
        }

        

        public string ParName(int index)
        {
            return index switch
            {
                0 => "List",
                1 => "Separator",
                _ => ""
            };
        }
    }
}
